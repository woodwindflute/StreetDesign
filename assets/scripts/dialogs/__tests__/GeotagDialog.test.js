/* eslint-env jest */
import React from 'react'
import { screen } from '@testing-library/react'
import { render } from '../../../../test/helpers/render'
import GeotagDialog from '../GeotagDialog'
import { isOwnedByCurrentUser } from '../../streets/owner'

// Mock dependencies that could break tests
jest.mock('../../streets/owner', () => ({
  isOwnedByCurrentUser: jest.fn()
}))

// as mocked, the intial state is a street with a previously saved location, and a new marker location
// in practice, these tests don't really handle this 'new' marker location yet

const initialState = {
  street: {
    creatorId: 'mayorofnullisland',
    location: {
      label: 'Null Island',
      wofId: '1234',
      latlng: {
        lat: 0,
        lng: 0
      }
    }
  },
  map: {
    markerLocation: { lat: -1, lng: -1 },
    addressInformation: {
      street: 'Null Island Port',
      label: 'Null Island Port, Null Island',
      id: 'polyline123'
    }
  },
  user: {
    geolocation: {
      data: {
        lat: 10,
        lng: 10
      }
    }
  },
  system: {
    devicePixelRatio: 1
  }
}

describe('GeotagDialog', () => {
  it('renders', () => {
    const { asFragment } = render(<GeotagDialog renderPopup={true} />, {
      initialState
    })
    expect(asFragment()).toMatchSnapshot()
  })

  it('allows a location to be cleared when the street is owned by the current user', () => {
    isOwnedByCurrentUser.mockReturnValueOnce(true)
    render(<GeotagDialog />, {
      initialState
    })

    expect(
      screen.queryByRole('button', { name: 'Clear location' })
    ).toBeInTheDocument()
  })

  it('allows a location to be added when the current user started this street', () => {
    isOwnedByCurrentUser.mockReturnValueOnce(true)
    render(<GeotagDialog />, {
      initialState: {
        ...initialState,
        street: {
          creatorId: 'mayorofnullisland',
          location: null
        }
      }
    })
    expect(
      screen.queryByRole('button', { name: 'Confirm location' })
    ).toBeInTheDocument()
  })

  /* neither confirm or clear location buttons should show up in this case */
  it('does not allow a location to be edited when the current user is not the street owner', () => {
    isOwnedByCurrentUser.mockReturnValueOnce(false)
    render(<GeotagDialog />, {
      initialState
    })
    // in this case we want to make sure neither button shows up, so we use regex to check the button name
    expect(
      screen.queryByRole('button', { name: /Confirm location|Clear location/ })
    ).not.toBeInTheDocument()
  })

  it('allows a location to be confirmed when the current anonymous user is not the street owner but there is no existing location attached', () => {
    isOwnedByCurrentUser.mockReturnValueOnce(false)
    render(<GeotagDialog />, {
      initialState: {
        ...initialState,
        street: {
          creatorId: 'creatorMadeThisWithNoLocation',
          location: null
        }
      }
    })

    expect(
      screen.queryByRole('button', { name: 'Confirm location' })
    ).toBeInTheDocument()
  })
  it.todo('does not show error banner if geocoding services are available')
  it.todo('cripples dialog behavior if geocoding services are unavailable')
})
