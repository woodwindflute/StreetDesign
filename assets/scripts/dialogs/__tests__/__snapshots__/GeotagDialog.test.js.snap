// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeotagDialog renders 1`] = `
<DocumentFragment>
  <div
    class="dialog-box-container"
  >
    <div
      class="dialog-box-backdrop"
    />
    <div
      class="dialog-box"
      role="dialog"
    >
      <button
        class="close"
        title="Dismiss"
      >
        <svg
          class="octicon"
        />
      </button>
      <div
        class="geotag-dialog"
      >
        <div
          class="geotag-input-container"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-0-label"
            class="geotag-input-form"
            role="combobox"
          >
            <input
              aria-autocomplete="list"
              aria-labelledby="downshift-0-label"
              autocomplete="off"
              class="geotag-input"
              id="downshift-0-input"
              placeholder="Search for a location"
              value=""
            />
          </div>
        </div>
        <div
          class="leaflet-container leaflet-grab leaflet-touch-drag"
          style="position: relative;"
          tabindex="0"
        >
          <div
            class="leaflet-pane leaflet-map-pane"
            style="left: 0px; top: 0px;"
          >
            <div
              class="leaflet-pane leaflet-tile-pane"
            >
              <div
                class="leaflet-layer "
                style="z-index: 1;"
              >
                <div
                  class="leaflet-tile-container leaflet-zoom-animated"
                  style="z-index: 18; left: 0px; top: 0px;"
                >
                  <img
                    alt=""
                    class="leaflet-tile"
                    role="presentation"
                    src="https://c.basemaps.cartocdn.com/rastertiles/voyager/12/2047/2047.png"
                    style="width: 256px; height: 256px; left: -251px; top: -221px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="leaflet-pane leaflet-shadow-pane"
            >
              <img
                alt=""
                class="leaflet-marker-shadow leaflet-zoom-hide"
                src="/images/marker-shadow.png"
                style="margin-left: -12px; margin-top: -41px; width: 41px; height: 41px; left: 5px; top: 35px;"
              />
            </div>
            <div
              class="leaflet-pane leaflet-overlay-pane"
            />
            <div
              class="leaflet-pane leaflet-marker-pane"
            >
              <img
                alt=""
                class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive leaflet-marker-draggable"
                src="/images/marker-icon.png"
                style="margin-left: -12px; margin-top: -41px; width: 25px; height: 41px; left: 5px; top: 35px; z-index: 35;"
                tabindex="0"
              />
            </div>
            <div
              class="leaflet-pane leaflet-tooltip-pane"
            />
            <div
              class="leaflet-pane leaflet-popup-pane"
            >
              <div
                class="leaflet-popup  leaflet-zoom-animated"
                style="bottom: -5px; left: 5px;"
              >
                <div
                  class="leaflet-popup-content-wrapper"
                >
                  <div
                    class="leaflet-popup-content"
                    style="width: 51px;"
                  >
                    <div
                      class="geotag-location-label"
                    >
                      Null Island
                    </div>
                  </div>
                </div>
                <div
                  class="leaflet-popup-tip-container"
                >
                  <div
                    class="leaflet-popup-tip"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="leaflet-control-container"
          >
            <div
              class="leaflet-top leaflet-left"
            >
              <div
                class="leaflet-control-zoom leaflet-bar leaflet-control"
              >
                <a
                  aria-label="Zoom in"
                  class="leaflet-control-zoom-in"
                  href="#"
                  role="button"
                  title="Zoom in"
                >
                  +
                </a>
                <a
                  aria-label="Zoom out"
                  class="leaflet-control-zoom-out"
                  href="#"
                  role="button"
                  title="Zoom out"
                >
                  −
                </a>
              </div>
            </div>
            <div
              class="leaflet-top leaflet-right"
            />
            <div
              class="leaflet-bottom leaflet-left"
            />
            <div
              class="leaflet-bottom leaflet-right"
            >
              <div
                class="leaflet-control-attribution leaflet-control"
              >
                <a
                  href="https://leafletjs.com"
                  title="A JS library for interactive maps"
                >
                  Leaflet
                </a>
                 | © 
                <a
                  href="https://www.openstreetmap.org/copyright"
                >
                  OpenStreetMap
                </a>
                , © 
                <a
                  href="https://carto.com/attribution"
                >
                  CARTO
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
